plugin/etcd/plugin.go

```diff
@@ -0,0 +1,451 @@
+		if err != nil {
+			fmt.Printf("@R{\u2717 user  %s}\n", err)
+			fail = true
+		} else if s == "" {
+			fmt.Printf("@G{\u2713 user} username was not provided so cert based auth will be used\n")
+		} else {
+			fmt.Printf("@G{\u2713 username} @C{%s}\n", plugin.Redact(s))
+		}
+
+		s, err = endpoint.StringValueDefault("pass", "")
+		if err != nil {
+			fmt.Printf("@R{\u2717 pass  %s}\n", err)
+			fail = true
+		} else if s == "" {
+			fmt.Printf("@R{\u2713 pass} password was not provided so cert based auth will be used\n")
+		} else {
+			fmt.Printf("@G{\u2713 password} @C{%s}\n", plugin.Redact(s))
+		}
+
+		s, err = endpoint.StringValueDefault("clientCertPath", "")
+		if err != nil {
+			fmt.Printf("@R{\u2717 client certificate path  %s}\n", err)
+		} else if s == "" {
+			fmt.Printf("@R{\u2713 client certificate path} was not provided\n")
+		} else {
+			fmt.Printf("@G{\u2713 client certificate path} was provided\n")
+		}
+
+		s, err = endpoint.StringValueDefault("clientKeyPath", "")
+		if err != nil {
+			fmt.Printf("@R{\u2717 client key path  %s}\n", err)
+		} else if s == "" {
+			fmt.Printf("@R{\u2713 client key path} was not provided\n")
+		} else {
+			fmt.Printf("@G{\u2713 client key path} was provided\n")
+		}
+
+		s, err = endpoint.StringValueDefault("caCertPath", "")
+		if err != nil {
+			fmt.Printf("@R{\u2717 CA certificate path  %s}\n", err)
+		} else if s == "" {
+			fmt.Printf("@R{\u2713 CA certificate path} was not provided\n")
+		} else {
+			fmt.Printf("@G{\u2713 CA certificate path} was provided\n")
+		}
+	}
+
+	b, err = endpoint.BooleanValueDefault("fullOverwrite", false)
+	if err != nil {
+		fmt.Printf("@R{\u2717 full restore  %s}\n", err)
+		fail = true
+	} else if b {
+		fmt.Printf("@G{\u2713} full restore enabled\n")
+	} else {
+		fmt.Printf("@G{\u2713} full restore disabled\n")
+	}
+
+	s, err = endpoint.StringValueDefault("prefix", "")
+	if err != nil {
+		fmt.Printf("@R{\u2717 prefix  %s}\n", err)
+		fail = true
+	} else if s == "" {
+		fmt.Printf("@R{\u2713 prefix} prefix was not provided so everything will be backed-up/restored\n")
+	} else {
+		fmt.Printf("@G{\u2713 prefix} @C{%s}\n", plugin.Redact(s))
+	}
+
+	if fail {
+		return fmt.Errorf("etcd: invalid configuration")
+	}
+
+	return nil
+}
+
+// Backup ETCD data
+func (p EtcdPlugin) Backup(endpoint plugin.ShieldEndpoint) error {
+	etcd, err := getEtcdConfig(endpoint)
+	if err != nil {
+		return err
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      etcd.ClientCertPath,
+		KeyFile:       etcd.ClientKeyPath,
+		TrustedCAFile: etcd.CaCertPath,
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{etcd.EtcdEndpoints},
+		DialTimeout: 2 * time.Second,
+		Username:    etcd.Username,
+		Password:    etcd.Password,
+		TLS:         tlsConfig,
+	})
```

Comment:
Most of the preceding code can be moved into a method on `*EtcdConfig`, and eliminate code duplication.